### PR TITLE
Clarified handling of rich dependencies and unified package string parsing.

### PR DESCRIPTION
**Pull request by PawelWMS**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [X] The toolchain has been rebuilt successfully (or no changes were made to it)
- [X] The toolchain/worker package manifests are up-to-date
- [X] Any updated packages successfully build (or no packages were changed)
- [X] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [X] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [X] All package sources are available
- [X] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [X] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [X] All source files have up-to-date hashes in the `*.signatures.json` files
- [X] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [X] Documentation has been updated to match any changes to the build system
- [ ] Ready to merge

---

###### Summary 

We've been parsing package strings with version like `gcc >= 1.2.3` with different logic inside `specreader.go` and `pkgjson.go`. The version from `specreader.go` failed to properly handle cases with missing spaces like `gcc>=1.2.3`. Since these instances are not consider erroneous by `rpmbuild` and we handle them correctly with the `PackagesListEntryToPackageVer` function from `pkgjson.go`, I've made `specreader.go` call into `PackagesListEntryToPackageVer` as well.

### Why the change
The unhandled `gcc>=1.2.3` cases in `specreader.go` made us create `PkgVer` structs with their `Name` fields set to `gcc>=1.2.3` and so far we only lucked out that this didn't cause any issues, as only one spec seems to be defining dependencies this way and it's not depending on another local spec.

### Updated handling of rich dependencies
The old way of splitting the dependencies by looking only for the ` or ` string was erroneous where it would attempt to create a `PackageVer` with a `Condition` field equal to `if` and `Version` field equal to `python3` for dependencies like `(python3-importlib-metadata if python3 < 3.8)` (taken from `python-tox.spec`).

So far, we ended having to handle rich dependencies from only two specs: `maven-surefire.spec` and `annobin.spec`. By accident we've managed to avoid build breaks in these cases.

The change from this PR makes the parsing logic and logging more robust.

I used the ["Boolean dependencies" paragraph](https://weldr.io/RPM-Dependencies/#boolean-dependencies) from an article by David Shea for guidance. Thank you, @dashea!

###### Change Log  

- Switched package strings parsing logic from `specreader.go` to use `PackagesListEntryToPackageVer` from `pkgjson.go`.
- Renamed `PackagesListEntryToPackageVer` to `PackageStringToPackageVer`.

###### Does this affect the toolchain?  

No.

###### Test Methodology

- [Full pipeline build](https://dev.azure.com/mariner-org/mariner/_build/results?buildId=406301&view=results).
- Local tests.
- Tooling unit tests.

### CODE REVIEW

File: toolkit/docs/how_it_works/3_package_building.md [hunk 1]
```diff
@@ -55,11 +55,14 @@ The single SPEC provides the base package `example`, but also provides a virtual
 
 All package dependency information is written to `./../build/pkg_artifacts/specs.json`.
 
-### Or Clauses
-Spec files can have `(a or b)` style requirements. When the build system encounters such a requirement it will record both options into the graph so that all possible requirements will be made available to pick from during package install allowing for maximum flexibility. This means that the build system requires all optional RPMs to be available to build/download even if they will not be used for a specific configuration.
+### Rich dependencies
+Spec files can have `(a <condition> b)` style requirements. Depending on the condition the following will happen::
```
Comment (cwize1) [inline toolkit/docs/how_it_works/3_package_building.md:59]: nit: Remove second `:`

Comment (PawelWMS) [reply toolkit/docs/how_it_works/3_package_building.md]: Thanks!

File: toolkit/tools/specreader/specreader.go [hunk 2]
```diff
@@ -498,40 +522,21 @@ func parseProvides(rpmsDir, toolchainDir string, toolchainRPMs []string, srpmPat
...
 // identifying if the package name is an "or" condition and returning all options.
-func parsePackageVersions(packagename string) (newpkgs []*pkgjson.PackageVer, err error) {
-	const (
-		NameField      = iota
-		ConditionField = iota
-		VersionField   = iota
-	)
+func parsePackageVersions(packageString string) (newPackages []*pkgjson.PackageVer, err error) {
+	// If the first character of the packageString is a "(" then it's an "or" condition.
+	if packageString[0] == '(' {
```
Comment (dmcilvaney) [inline toolkit/tools/specreader/specreader.go:530]: I think `parsePackageVersions()` was the only questionable caller that may have passed bad input, but we might want to run a trim on the input if we are going to be checking the 1st character for `'('`.

File: toolkit/tools/specreader/specreader.go [hunk 3]
```diff
@@ -587,26 +592,60 @@ func condensePackageVersionArray(packagelist []*pkgjson.PackageVer, specfile str
...
-	packagename = strings.ReplaceAll(packagename, "(", "")
-	packagename = strings.ReplaceAll(packagename, ")", "")
+// parseRichDependency splits a package name like '(foo or bar)' and returns both foo and bar as separate requirements.
+func parseRichDependency(richDependency string) (versions []*pkgjson.PackageVer, err error) {
+	const documentationHint = "Please refer to 'docs/how_it_works/3_package_building.md#rich-dependencies' for explanation of limitations"
 
-	packageSplit := strings.Split(packagename, " or ")
-	err = minSliceLength(packageSplit, 1)
+	for _, singleCondition := range unsupportedBooleanConditions {
+		if strings.Contains(richDependency, singleCondition) {
```
Comment (dmcilvaney) [inline toolkit/tools/specreader/specreader.go:600]:
```suggestion
		// 'singleCondition' is of the form " <unsupported_condition> " and will match whole-word operators only.
		if strings.Contains(richDependency, singleCondition) {
```

**Review by dmcilvaney — APPROVED** [review]

**Pull request merged**